spec/index.html

```diff
@@ -1493,3 +1503,221 @@ <h1>SharedArray ( ..._values_ )</h1>
+    </emu-note>
+
+    <emu-clause id="sec-atomics.mutex.lock">
+      <h1>Atomics.Mutex.lock ( _mutex_ [ , _unlockToken_ [ , _timeout_ ] ] )</h1>
+      <p>This function puts the surrounding agent in a wait queue and suspends it until the mutex is unlocked or until the wait times out, returning a String differentiating those cases.</p>
+      <p>It performs the following steps when called:</p>
+      <emu-alg>
+        1. If _mutex_ is not an Object, throw a *TypeError* exception.
+        1. If _mutex_ does not have a [[MutexWaiterList]] internal slot, throw a *TypeError* exception.
+        1. If _unlockToken_ not *undefined*, then
+          1. If _unlockToken_ is not an Object, throw a *TypeError* exception.
+          1. If _unlockToken_ does not have a [[LockedMutex]] internal slot, throw a *TypeError* exception.
+          1. If _unlockToken_.[[LockedMutex]] is not ~empty~, throw a *TypeError* exception.
+        1. If _timeout_ is *undefined*, set _timeout_ to *+∞*<sub>𝔽</sub>.
+        1. If _timeout_ is not a Number, throw a *TypeError* exception.
+        1. If _timeout_ is either *NaN* or *+∞*<sub>𝔽</sub>, let _t_ be +∞; else if _timeout_ is *-∞*<sub>𝔽</sub>, let _t_ be 0; else let _t_ be max(ℝ(_timeout_), 0).
+        1. If AgentCanSuspend() is *false*, throw a *TypeError* exception.
+        1. Let _result_ be LockMutex(_mutex_, _t_).
+        1. If _result_ is ~acquired~, then
+          1. Return UnlockTokenCreateIfNeeded(_unlockToken_, _mutex_).
+        1. Else,
+          1. Assert: _result_ is ~timed-out~.
+          1. Return *"timed-out"*.
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-atomics.mutex.trylock">
+      <h1>Atomics.Mutex.tryLock ( _mutex_ [ , _unlockToken_ ] )</h1>
+      <p>It performs the following steps when called:</p>
+      <emu-alg>
+        1. If _mutex_ is not an Object, throw a *TypeError* exception.
+        1. If _mutex_ does not have a [[MutexWaiterList]] internal slot, throw a *TypeError* exception.
+        1. If _unlockToken_ not *undefined*, then
+          1. If _unlockToken_ is not an Object, throw a *TypeError* exception.
+          1. If _unlockToken_ does not have a [[LockedMutex]] internal slot, throw a *TypeError* exception.
+          1. If _unlockToken_.[[LockedMutex]] is not ~empty~, throw a *TypeError* exception.
+        1. Let _WL_ be _mutex_.[[MutexWaiterList]].
+        1. Perform EnterCriticalSection(_WL_).
+        1. If _mutex_.[[IsLocked]] is *false*, then
+          1. Set _mutex_.[[IsLocked]] to *true*.
+          1. Let _result_ be UnlockTokenCreateIfNeeded(_unlockToken_, _mutex_).
+        1. Else,
+          1. Let _result_ be *undefined*.
+        1. Perform LeaveCriticalSection(_WL_).
+        1. Return _result_.
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-atomics.mutex.withlock">
+      <h1>Atomics.Mutex.withLock ( _mutex_, _callback_ [ , _thisArg_ ] )</h1>
```

Comment:
Do we want `tryWithLock`? Not sure what to return in the already-locked case, though. Maybe have a parameter to use for that value?

Reply:
I don't see that often in other PLs/stdlibs. I'm not against it though.

Reply:
Swift 6 (new as of 2 days ago) has [withLockIfAvailable](https://developer.apple.com/documentation/synchronization/mutex/withlockifavailable(_:))[^1], which is this. I also [think that's a better name](https://github.com/tc39/proposal-structs/issues/29) than `tryLock` / `tryWithLock`, incidentally.

Really it's just filling out the missing quadrant on the (lock()+unlock() vs callback) x (blocking vs only-if-available) matrix, though, so it's pretty natural if we already have the other three quadrants filled out.

[^1]: You can read through its history [in this thread](https://forums.swift.org/t/pitch-synchronous-mutual-exclusion-lock/69889/56) if you want to.

Reply:
> In the tryWithLock case we'd probably want to return a value indicating the lock was already owned rather than invoking the callback to maintain symmetry.

I didn't add `tryWithLock` (whatever its name) because I couldn't figure out how to do this. I suppose we can return a distinguishing value if we always discard the return value of the callback, instead of trying to return that.

Reply:
> I didn't add `tryWithLock` (whatever its name) because I couldn't figure out how to do this. I suppose we can return a distinguishing value if we always discard the return value of the callback, instead of trying to return that.

Isn't that what `Mutex.tryLock` did in the dev trial? It always returned a boolean, so if you needed a value from the callback you would have to use a closure.

Reply:
Yeah, but I got some feedback that was annoying.

Reply:
This is a case where returning an object with a `{ success, result }` object might make sense. If we have a lock-token producing `tryLock`, then maybe a callback version isn't quite as necessary? If you are conditionally locking then you are already buying into more complexity anyways.

Reply:
I've waffled on this and think we shouldn't propose `withLock` for the initial proposal, as the unlock token design means the callback-taking approach doesn't add that much value. My thinking is the main value of `withLock` is you don't have to think about locking or unlocking operations, but with unlock tokens, the simple version doesn't compose well with condition variables. If you want to wait for a condition inside the callback, that means the callback needs to be passed an unlock token.